ibis/backends/clickhouse/compiler/core.py

```diff
@@ -68,31 +82,80 @@ def translate(op: ops.TableNode, params: Mapping[ir.Value, Any]) -> sg.exp.Expre
     sqlglot.expressions.Expression
         A sqlglot expression
     """
-    params = {param.op(): value for param, value in params.items()}
 
     alias_index = 0
     aliases = {}
 
-    def fn(node, cache, params=params, **kwargs):
+    def fn(node, _, **kwargs):
         nonlocal alias_index
 
+        result = _translate_node(node, aliases=aliases, **kwargs)
+
+        if not isinstance(node, ops.TableNode):
+            return result
+
         # don't alias the root node
         if node is not op:
             aliases[node] = f"t{alias_index:d}"
             alias_index += 1
 
-        raw_rel = translate_rel(
-            node, aliases=aliases, params=params, cache=cache, **kwargs
-        )
-
         if alias := aliases.get(node):
             try:
-                return raw_rel.subquery(alias)
+                return result.subquery(alias=alias)
             except AttributeError:
-                return sg.alias(raw_rel, alias)
+                return sg.alias(result, alias=alias)
         else:
-            return raw_rel
-
-    results = op.map(fn, filter=ops.TableNode)
+            return result
+
+    # substitute parameters immediately to avoid having to define a
+    # ScalarParameter translation rule
+    #
+    # this lets us avoid threading `params` through every `translate_val` call
+    # only to be used in the one place it would be needed: the ScalarParameter
+    # `translate_val` rule
+    params = {param.op(): value for param, value in params.items()}
+    replace_literals = p.ScalarParameter >> (
+        lambda op, _: ops.Literal(value=params[op], dtype=op.dtype)
+    )
+
+    # rewrite cumulative functions to window functions, so that we don't have
+    # to think about handling them in the compiler, we need only compile window
+    # functions
+    replace_cumulative_ops = p.WindowFunction(
+        x @ p.Cumulative, y
+    ) >> a.cumulative_to_window(x, y)
+
+    # replace the right side of InColumn into a scalar subquery for sql
+    # backends
+    replace_in_column_with_table_array_view = p.InColumn >> (
+        lambda op, _: op.__class__(
+            op.value,
+            ops.TableArrayView(
+                ops.Selection(
+                    table=an.find_first_base_table(op.options), selections=(op.options,)
```

Comment:
We should switch to `Topmost` pattern to locate the base table, but in a follow-up.

Reply:
Something like `BaseTableOf = Topmost(ops.Relation)`